lab-dustin/server.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const net = require('net');
+const PORT = process.env.PORT || 3000;
+const server = net.createServer();
+const constructor = require('./client-constructor');
+
+const EE = require('events');
+const ee = new EE();
+
+const pool = [];
+
+ee.on('@all',function(sender,message){
+  pool.forEach(receiver => {
+    if (receiver.id === sender.id) return;
+
+    receiver.socket.write(`${sender.id}: ${message}\r\n`);
+  });
+});
+
+ee.on('@username',function(sender,name){
+  console.log(name,sender.id);
+  var usernameExists = false;
+  pool.forEach(function(user){
+    if (user.id === name){
+      sender.socket.write('Username already exists.\r\n');
+      usernameExists = true;
```

Comment:
Fun thought experiment: what would calculating `usernameExists` look like as a `reduce`?

Really, though, the best higher-order function to use here is [`Array.prototype.some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).